File: Solutions/Combination_Sum_III/Combination_Sum_III.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 
 combination no repeating # []
 
 combination is unique
 
 n === target
 
 k is numbers
 
 1 - 9 
 
 */
var combinationSum3 = function(k, n) {
    let res = [];
    let dfs = function(a, sum, i) {
        if (a.length === k) {
            if (sum === n) res.push([...a]);
            return;
        }
        
        for (i; i <= 9; i++) {
            a.push(i);
            dfs(a, sum + i, i+1);
            a.pop();
        }
        
    }
    dfs([], 0, 1);
    return res;
};

//Fastest
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
    const candidates = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    const res = [];
    helper(res, candidates, k, n);
    return res;
};

function helper(res, candidates, maxLength, target, path = [], idx = 0) {
    if (target < 0) return;
    if (path.length > maxLength) return;
    if (target === 0 && path.length === maxLength) {
        res.push(path);
        return;
    }
    for (let i = idx; i < candidates.length; i++) {
        const curr = candidates[i];
        const newTarget = target - curr;
        helper(res, candidates, maxLength, newTarget, [...path, curr], i + 1);
    }
}

File: Solutions/Combination_Sum_III/Combination_Sum_III.py
#Most commmon
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]: 
        ans = []
        self.dfs(k, n, ans, [], 1)
        return ans
    
    def dfs(self, k, n, ans, rslt, i):
        if len(rslt) == k:
            if sum(rslt) == n:
                ans.append(rslt)
            return

        for j in range(i, 10):
            self.dfs(k, n, ans, rslt + [j], j + 1)

#Fastest
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        
        res = []
        self.dfs(range(1,10), k, n, 0, [], res)
        return res

    def dfs(self, nums, k, n, index, path, res):
        if k < 0 or n < 0: # backtracking 
            return 
        if k == 0 and n == 0: 
            res.append(path)
        for i in range(index, len(nums)):
            self.dfs(nums, k-1, n-nums[i], i+1, path+[nums[i]], res)
